js/ui/status/missionGame.js

```diff
@@ -0,0 +1,528 @@
+                lines[lineIndex] = lines[lineIndex] + append.replace(/\n/g, "");;
+            }
+
+            /* If we encountered a '\n', we also need to
+             * break lines, though this will just
+             * involve inserting a new line. */
+            if (text[textCounter] === '\n') {
+                lines.push(prefix);
+                lineIndex++;
+                lineCounter = prefix.length;
+            }
+
+            lineCounter += append.length;
+            lastSpace = textCounter;
+        }
+    }
+
+    return lines;
+}
+
+function copyObject(src) {
+    let dst = {};
+    Lang.copyProperties(src, dst);
+    return dst;
+}
+
+const ScrolledLabel = new Lang.Class({
+    Name: 'ScrolledLabel',
+    Extends: St.Label,
+
+    _init: function(params) {
+        let parentParams = copyObject(params);
+        parentParams.text = '';
+
+        this.parent(parentParams);
+        this._text = params.text;
+        this._textIndex = 0;
+        this._scrollTimer = 0;
+        this._waitCount = 0;
+        this.complete = false;
+
+        this.style_class = 'chatbox-character-text';
+
+        this.set_x_expand(true);
+        this.clutter_text.set_x_expand(true);
+    },
+    start: function(scrollView) {
+        /* Avoid double-start */
+        if (this._scrollTimer) {
+            return;
+        }
+
+        /* Immediately display the first character so that
+         * scrolling to the bottom will work */
+        this._textIndex++;
+        this.set_text(this._text.slice(0, this._textIndex));
+
+        /* Add a timeout to gradually display all this text */
+        this._scrollTimer = GLib.timeout_add(GLib.PRIORITY_DEFAULT, 10, Lang.bind(this, function() {
+            if (this._waitCount > 0) {
+                this._waitCount--;
+                return true;
+            }
+
+            if (this._textIndex === this._text.length) {
+                this.complete = true;
+                this.emit('finished-scrolled');
+                return false;
+            }
+
+            this._textIndex++;
+            this.set_text(this._text.slice(0, this._textIndex));
+
+            /* Stop on punctuation */
+            if ("!?.".indexOf(this._text[this._textIndex - 1]) !== -1) {
+                this._waitCount = 100;
+                return true;
+            }
+
+            /* Running this on a timer every time is not
+             * ideal, but I haven't found a better way
+             * ensure the view is reliably always
+             * scrolled */
+            scrollView();
+
+            return true;
+        }));
+    },
+    fastForward: function() {
+        this._textIndex = this._text.length - 1;
+    }
+});
+Signals.addSignalMethods(ScrolledLabel.prototype);
+
+WrappedLabel = new Lang.Class({
+    Name: 'WrappedLabel',
+    Extends: ScrolledLabel,
+    _init: function(params) {
+        this.parent(params);
+        this._text = wrapTextWith(this._text, WRAP_CONSTANT, "> ").join("\n");
```

Comment:
You can pass the text in the parameters instead of setting a private field from outside like this.